New Witness struct to improve ser/de perfomance

RCasatta opened the pull request:
At the moment the Witness struct is  `Vec<Vec<u8>>`, the vec inside a vec cause a lot of allocations, specifically:

- empty witness -> 1 allocation, while an empty vec doesn't allocate, the outer vec is not empty
- witness with n elements -> n+1 allocations

The proposed Witness struct contains the serialized format of the witness. This reduces the allocations to:

- empty witness -> 0 allocations
- witness with n elements -> 1 allocation for most common cases (you don't know how many bytes is long the entire witness beforehand, thus you need to estimate a good value, not too big to avoid wasting space and not too low to avoid vector reallocation, I used 128 since it covers about 80% of cases on mainnet)

The inconvenience is having slightly less comfortable access to the witness, but the iterator is efficient (no allocations) and you can always collect the iteration to have a Vec of slices. If you collect the iteration you end up doing allocation anyway, but the rationale is that it is an operation you need to do rarely while ser/de is done much more often.

I had to add a bigger block to better see the improvement (ae860247e191e2136d7c87382f78c96e0908d700), these are the results of the benches on my machine:

```
RCasatta/master_with_block
test blockdata::block::benches::bench_block_deserialize                 ... bench:   5,496,821 ns/iter (+/- 298,859)
test blockdata::block::benches::bench_block_serialize                   ... bench:     437,389 ns/iter (+/- 31,576)
test blockdata::block::benches::bench_block_serialize_logic             ... bench:     108,759 ns/iter (+/- 5,807)
test blockdata::transaction::benches::bench_transaction_deserialize     ... bench:         670 ns/iter (+/- 49)
test blockdata::transaction::benches::bench_transaction_get_size        ... bench:           7 ns/iter (+/- 0)
test blockdata::transaction::benches::bench_transaction_serialize       ... bench:          51 ns/iter (+/- 5)
test blockdata::transaction::benches::bench_transaction_serialize_logic ... bench:          13 ns/iter (+/- 0)

branch witness_with_block (this one)
test blockdata::block::benches::bench_block_deserialize                 ... bench:   4,302,788 ns/iter (+/- 424,806)
test blockdata::block::benches::bench_block_serialize                   ... bench:     366,493 ns/iter (+/- 42,216)
test blockdata::block::benches::bench_block_serialize_logic             ... bench:      84,646 ns/iter (+/- 7,366)
test blockdata::transaction::benches::bench_transaction_deserialize     ... bench:         648 ns/iter (+/- 77)
test blockdata::transaction::benches::bench_transaction_get_size        ... bench:           7 ns/iter (+/- 0)
test blockdata::transaction::benches::bench_transaction_serialize       ... bench:          50 ns/iter (+/- 5)
test blockdata::transaction::benches::bench_transaction_serialize_logic ... bench:          14 ns/iter (+/- 0)
```

With an increased performance to deserialize a block of about 21% and to serialize a block of about 16% (seems even higher than expected, need to do more tests to confirm, I'll appreciate tests results from reviewers)

RCasatta commented on src/util/bip143.rs:
```diff
@@ -177,10 +176,10 @@ impl<R: DerefMut<Target=Transaction>> SigHashCache<R> {
     ///     let _sighash = sig_hasher.signature_hash(inp, &prevout_script, 42, SigHashType::All);
     ///     // ... sign the sighash
-    ///     sig_hasher.access_witness(inp).push(Vec::new());
+    ///     //sig_hasher.access_witness(inp).push(Vec::new());
```
I still have to implement `push` on Witness...

dr-orlovsky commented:
Have you seen https://github.com/rust-bitcoin/rust-bitcoin/pull/668/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR749-R825 ? It should be very complimentary to this PR

RCasatta commented:
> Have you seen https://github.com/rust-bitcoin/rust-bitcoin/pull/668/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR749-R825 ? It should be very complimentary to this PR

I didn't but I've looked it now, I am not sure to grasp the complementarity since one it's address witness and the other is the witness stack in the inputs, could you elaborate?

dr-orlovsky commented:
By "complementing" I mean they solve strong types issue for segwit related structures. I gave you link just to prevent double work from you if you will occasionally do that type which I already did

TheBlueMatt commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,209 @@
+
+    /// contains the witness Vec<Vec<u8>> serialization without the initial varint indicating the
+    /// number of elements (which is stored in len)
+    content: Vec<u8>,
```
If we're going down this path, does it make sense to also consider something like Bitcoin Core's "prevec" for the witness contents?

elichai replied:
FWIW: https://docs.rs/tinyvec/1.5.0/tinyvec/enum.TinyVec.html

TheBlueMatt replied:
Yea, I mean such a thing is super-duper trivial to write.

RCasatta replied:
I think a tinyvec like structure could probably be applicable broadly (like also for inputs/outpus) and as an addition to this, probably better to address in another PR

TheBlueMatt commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,209 @@
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]  //TODO manually implement?
+pub struct Witness {
```
Does it make sense to have a "pointers to the start of each witness element" vec as well? I guess given most witnesses are like 2-5 elements long maybe its not worth it, but would be cool to add a benchmark for iterating a lightning HTLC transaction witness and just test.

RCasatta replied:
Pointers to the start of each witness element would allow efficient index access but require another Vec.
I think for witness is not worth it because, as you said usually the witness is a few elements.

Kixunil replied:
Maybe have offsets in [`SmallVec`](https://docs.rs/smallvec/1.7.0/smallvec/)?
We could also optimistically assume that transactions are smaller than 4M bytes, so we could cram offsets into `struct U24(u8, u8, u8);`

RCasatta replied:
> Maybe have offsets in SmallVec?
https://github.com/rust-bitcoin/rust-bitcoin/pull/672#discussion_r721120421 
Also, having offsets would allow efficient indexed access, but mutable access it's another story... In the end I don't think there are really use case requiring that. I think I would not add offsets here even with tinyvec/smallvec.

> We could also optimistically assume that transactions are smaller than 4M bytes, so we could cram offsets into struct U24(u8, u8, u8);

the overall weight of the struct would not change because of padding (on most common 64 bits arch), I would leave this optimization in case we need other fields in the struct

TheBlueMatt reviewed (commented):
This is awesome! I've historically always wanted to do this on a broader scale - think a similar "store it in serialized form" for an entire transaction/block and just store pointers to the start of various datastructures to provide accessors. Of course that could come later, but this is awesome by itself.

dpc commented:
> iterator is efficient (no allocations) and you can always collect the iteration to have a Vec of slices

One allocation improves cache locality so even ignoring allocations, it's faster to go through one chunk of linear memory than jump between many.

RCasatta commented:
Some updates:

- squashed commits
- added fuzzing
- fix overflows found with fuzzing
- documented a bit
- removed the big block binary used to bench (cherry-pick 8412926e7485373a32a2edd1ec844b0cd6b0593a in case you want to bench)
- checked some onchain witness usage, and apparently, 128 https://github.com/rust-bitcoin/rust-bitcoin/pull/672/commits/0c9300f4b114d861f49736bf1838437206d6809d#diff-13b19787ce1492ec1fe05a7e9cbba8bd24735ca71bb87e0e2f58f835b220a0feR76 is good enough to handle 80% of witnesses -> see https://opreturn.org/witness-stats/

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,327 @@
+//! Witness
+//!
+//! This module contains the [`Witness`] struct and relative methods to operate on it
```
Nit: I think you meant related?

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,327 @@
+/// Support structure to allow efficient and convenient iteration over the Witness elements
+pub struct WitnessIterator<'a> {
+    witness: &'a Witness,
+    cursor: usize,
```
Why not just `core::slice::Iter<'a, u8>`?

Kixunil replied:
You can, it has [`as_slice()`](https://doc.rust-lang.org/std/slice/struct.Iter.html#method.as_slice) method. I used it in #662 and #673

Was also thinking of doing [`take_slice_or_kill`](https://github.com/rust-bitcoin/rust-bitcoin/pull/673/files#diff-af6864f932193bfbb57ebf73cd8c41bc12a33a25e3b27fa47ce95c62392158daR649) via extension trait, so it could be reused. You seem to be doing something similar so maybe worth revisiting that idea.

(Was also considering adding it to `core` as it seems widely applicable.)

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,327 @@
+    }
+}
+
+impl<'a> Iterator for WitnessIterator<'a> {
```
Nit: I think the naming for iterator **structs** tends to end with `Iter`, not `Iterator` (e.g. `slice::Iter`).

Also see: https://rust-lang.github.io/api-guidelines/naming.html#iterator-type-names-match-the-methods-that-produce-them-c-iter-ty

Kixunil replied:
Yes, I believe so.

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,327 @@
+    where
+        S: serde::Serializer,
+    {
+        let vec: Vec<_> = self.iter().map(|e| e.to_vec()).collect();
```
I think this could be faster but for initial code probably good. (Manual serde impls are annoying as hell.)

Kixunil commented:
@dpc yes, also `bytes` does something similar. I wanted to avoid doing a bunch of atomic operations for every transaction.

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -198,17 +192,15 @@ impl<'a> Iterator for Iter<'a> {
-            self.cursor = end;
-            Some(&vec[start..end])
+        let varint = VarInt::consensus_decode(self.0.as_slice()).ok()?;
+        self.0.nth(varint.len() - 1)?; // VarInt::len returns at least 1
```
I'd like to see `VarInt::consensus_decode_iter()` one day but good enough for this PR.

dr-orlovsky commented:
Needs a rebase

dr-orlovsky reviewed (changes requested):
It is also required to change type of the witness data in PSBT inputs

dr-orlovsky commented:
Can you also pls stick in `Display` implementation:

```rust
impl Display for Witness {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        for item in self.iter() {
            writeln!(f, "{}", item.to_hex())?;
        }
        Ok(())
    }
}
```

dr-orlovsky commented:
Here are all the improvements I'd like to have in this PR: https://github.com/RCasatta/rust-bitcoin/pull/4

RCasatta commented:
> Without acessor methods to items inside Witness (Index impl or get) it is literally impossible to use this structure in many cases (like in miniscript etc)

I don't think it's literally impossible to access elements without `Index`, the options are:
* calling `nth` on a new iterator (which is basically what the suggested https://github.com/RCasatta/rust-bitcoin/pull/4/commits/bf3127d856f65b8dc6f847e5446e531e05c8d6aa is doing)
* collect the iterator and use the `Vec<Vec<u8>>`

Also, I can't find extensive indexed access in the miniscript lib (please point me to the code if I am wrong).

So I am not convinced to implement `Index` on this Witness type which could bring to inefficient code like
```
for i in 0..witness_len {
   do something with witness[i]
}
```

I am open to changing my mind, what do other people think?

Kixunil commented:
Why `Display`? Sounds strange to display it. `Debug` definitely yes.

dpc commented:
> I am open to changing my mind, what do other people think?

This PR seems to be skipping deserializing witness, so not everyone has to pay for it, and code that do need it, can pay for it how they want. Seems like a win in most cases, and a small inconvenience for other.

dr-orlovsky commented:
> Why Display? Sounds strange to display it.

Well, I'd like to have in command-line pro wallets (I have one myself) printing tx witness like
```
<stack-elem-1>
<stack-elem-2>
...
```

And not

```
Witness {
	data: 
		135,
		76,
....
```

dr-orlovsky commented:
@RCasatta replied in that PR, but briefly, we need things like "accessing second to last element" in many cases, including parsing witness data for taproot and segwit v0, so why not to have a index map as a part of the structure and const-time `Index` impls?

RCasatta commented:
> @RCasatta replied in that PR, but briefly, we need things like "accessing second to last element" in many cases, including parsing witness data for taproot and segwit v0, so why not to have a index map as a part of the structure and const-time `Index` impls?

Adding an index map would defeat the main [1] purpose of this PR [2] because it would require another allocation for the map. 

If "accessing second to last element" is a common use case (and also the last, I guess), we can add two Option<NonZeroUsize> pointing to the last two elements and two efficient `last()` and `second_to_last()` methods

[1]: main because I think deserialization happens more than operating on the witness
[2]: speeding up deserialization

dr-orlovsky commented:
I think that is a good solution!

dpc commented:
If there is are multiple strong usecases here, how about having two (or even multiple `Witness`). There's some type system parametrization  pollution involved, but everyone can get what they desire. So any type implementing consensus decode and encode could be used (obviously it makes sense only for some actual types).

* This new `Witness` becomes `CompactWitness`,
* The `Vec<Vec<u8>>` becomes `DeepWitness`
* if you have any other ideas, you can create `AnotherWitness`
* `TxIn` becomes `TxIn<W = CompactWitness>`, and recursively any type holding witness gets a `<W = CompactWitness>` added. It is a bit of noise all over the implementation, but in a typical use one can still use `TxIn` and it defaults to `CompactWitness`.

TheBlueMatt commented:
> If "accessing second to last element" is a common use case (and also the last, I guess), we can add two Option<NonZeroUsize> pointing to the last two elements and two efficient last() and second_to_last() methods

Wouldn't this still kill performance of this PR? As an alternative, the simplest solution seems to be the one already proposed ("collect the iterator and use the Vec<Vec<u8>>"). I'd like to understand why that is unacceptable for miniscript/PSBT. If its not, is it possible to track the starting offsets in a second datastructure that wraps the transaction (and provides `Deref<Target=Transaction>`)? Explicitly tracking the offsets may be okay performance-wise since we have to hit them anyway during deserialization (ugh, bitcoin serialization formats....) but ideally of course we'd avoid the extra malloc.

dpc commented:
> How would this work - would we have to make `Transaction` generic across some `Witness` trait?

Yes `trait Witness: Encode + Decode {}` or something and it bubbles up from `TxIn` upwards, though with default generic type it would be usually rather invisible to the user code.

Kixunil commented:
IMO for advanced cases we should just have streaming (possibly zero-copy) deserailization. Then everyone who needs to deserailize fast can just write their own hooks without ton of maintenance burden for us. This may be useful for `electrs` synchronization actually (not sure how much it improves perf though).

TheBlueMatt commented:
> IMO for advanced cases we should just have streaming (possibly zero-copy) deserailization.

The really nice thing about this PR is it doesn't require a ton of downstream changes. Just swap a few vec accesses for an `iter().skip().next()`. Most users of `Transaction` don't need the full witness in a vec, and if they just iterate it once or twice this PR is still faster than putting the whole thing in `Vec<Vec<>>`, so its faster for probably 99% of use-cases, even if you have to iterate instead of direct-access.

I'm struggling to come up with a reason to keep the current code over this PR, honestly. Its not *quite* as nice for users, but its substantially faster for users. Doing some zero-copy deserialization, on the other hand, is a *big* change for users. Suddenly objects refer to the owning backing bytes and you can't (trivially) edit them in memory (so I guess we'd have two copies of every consensus object?).

Kixunil commented:
I meant zero-copy deserialization as an additional feature, not as replacement for the current system. The current deserialization may be reimplementable using zero-copy algorithm by just copying. :) I was reacting to @dpc suggestion to use generics. While I like generics, I think zero-copy would be more flexible.

Pseudo-code:

```rust
fn visit_output(&mut self, script: &'de [u8], amount: u64) {
    self.output.push(TxOut { script: Script::from(script.to_owned()), amount, });
}
```

> I'm struggling to come up with a reason to keep the current code over this PR

I agree this is better.

TheBlueMatt commented:
> I was reacting to @dpc suggestion to use generics.

Ah! Apologies, I misunderstood the context for your comment.

RCasatta commented:
> Needs a rebase

Rebased

> Wouldn't this still hurg performance of this PR (which is the whole point)?

No, because the added pointers are only 2, specific to access last and second_to_last elements, so they are on the stack and they don't require other allocations.
Cherry-picking ae860247e191e2136d7c87382f78c96e0908d700 to test with a big block we have: 

```
test blockdata::block::benches::bench_block_deserialize                 ... bench:   4,275,568 ns/iter (+/- 210,037)
test blockdata::block::benches::bench_block_serialize                   ... bench:     379,426 ns/iter (+/- 40,914)

```
which is similar to the best result I posted initially

RCasatta commented on src/blockdata/witness.rs:
```diff
@@ -27,23 +28,35 @@ pub struct Witness {
     /// It is stored separately (instead of as VarInt in the initial part of content) so that method
     /// like [`Witness::push`] doesn't have case requiring to shift the entire array
-    witness_elements: u64,
+    witness_elements: usize,
```
we may want to save 8 bytes in the Witness struct  using u32 for `witness_elements`, `last` and `second_to_last` and still support MAX_VECTOR_SIZE witnesses, however I thought using usize was a safer bet initially

sanket1729 replied:
`last` and `second_to_last` are not reset when calling the clear function

RCasatta commented:
Rebased after #702

dr-orlovsky reviewed (commented):
I started reviewing this some time ago but was interrupted. Posted my two pending comments from that attempt.

...and needs a rebase again

RCasatta commented on src/util/bip143.rs:
```diff
@@ -177,10 +176,10 @@ impl<R: DerefMut<Target=Transaction>> SigHashCache<R> {
     ///     let _sighash = sig_hasher.signature_hash(inp, &prevout_script, 42, SigHashType::All);
     ///     // ... sign the sighash
-    ///     sig_hasher.access_witness(inp).push(Vec::new());
+    ///     sig_hasher.access_witness(inp).push(&Vec::new());
```
done

RCasatta commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    /// Push a new element on the witness, require an allocation
+    pub fn push(&mut self, new_element: &[u8]) {
```
done

RCasatta commented:
Rebased

apoelstra commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,328 @@
+    cursor: usize,
+}
+
+impl From<Vec<Vec<u8>>> for Witness {
```
In c342adbb1de1adc3dfc52f1b1cf8116f1d37b049:

I think the `From` trait is inappropriate here because you're doing a new allocation and the conversion is nontrivial. Can you pull this into an independent `from_vec` or something method? (Also, could you add a `to_vec` method which is ust a convenience wrapper over iterate-and-collect?)

Relatedly, in your serde `Deserialize` impl, I'm sure we can do something more efficient than decoding a `Vec<Vec<u8>>` and then converting -- though for this PR that's fine, we can do optimizations later.

RCasatta replied:
I added 8d23404a268eba662ac97a3292cf7247ce974c5d
If the commit is what you expected I guess I'll squash the last 4 commits in the first

RCasatta replied:
Not going to squash since it received the ACK it looks it's fine like it is

apoelstra commented:
Done reviewing 00b2dd542637e545ef71e50667276c6e8f3c302b. Looks great to me, just one comment.

Thanks for adding the last and second-to-last accessors. I think these are pretty commonly useful, since in a complete witness stack the last element is the witness script, while the second-to-last is the actual "bottom" of the stack.

apoelstra reviewed (approved):
ACK 8d23404a268eba662ac97a3292cf7247ce974c5d

sanket1729 commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,328 @@
+        self.witness_elements = 0;
+    }
+
+    /// Push a new element on the witness, require an allocation
```
nit: requires an allocation.

RCasatta replied:
done

sanket1729 commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,328 @@
+        assert_eq!(witness_serialized, serialize(&witness));
+
+        //assert_eq!(32, std::mem::size_of::<Witness>());
+        //assert_eq!(24, std::mem::size_of::<Option<Vec<u8>>>());
```
nit: remove comments here?

RCasatta replied:
They were removed in latest commit, but present in old one, after the squash they are removed completely

sanket1729 reviewed (commented):
Did a detailed code review. Looks good, left one comment about clear function. 

I was going to ask for a pop function, but I think for verifying and interpreting scripts witnesses, we should convert it to Vec<Vec<u8>>.

RCasatta commented:
Fixed the clear method.

I squashed all but the fuzz commit into one since they were changing the same code.

sanket1729 reviewed (approved):
ACK 2fd0125bfa3248c860be2bb91805e3cc6416e692

apoelstra reviewed (approved):
ACK 106acdc3ac54372c4cc0a4a011a2ab8848159643

RCasatta commented:
> ACK 2fd0125bfa3248c860be2bb91805e3cc6416e692

Do you mean the tip commit?

RCasatta commented:
@dr-orlovsky could you remove the "changes requested" if everything is ok? I cannot merge otherwise

sanket1729 reviewed (approved):
ACK 106acdc3ac54372c4cc0a4a011a2ab8848159643

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -657,10 +656,10 @@ impl<R: DerefMut<Target = Transaction>> SigHashCache<R> {
     ///     let _sighash = sig_hasher.segwit_signature_hash(inp, &prevout_script, 42, EcdsaSigHashType::All);
     ///     // ... sign the sighash
-    ///     sig_hasher.witness_mut(inp).unwrap().push(Vec::new());
+    ///     sig_hasher.witness_mut(inp).unwrap().push(&Vec::new());
```
Nit:

```suggestion
    ///     sig_hasher.witness_mut(inp).unwrap().push(&[]);
```

dr-orlovsky commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,395 @@
+use prelude::*;
+use VarInt;
+
+#[cfg(feature = "serde")]
+use serde;
```
Nit: this seems redundant (we always have access to the crate without `use`ing it)

dr-orlovsky commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,395 @@
+
+impl Witness {
+    /// Creates [`Witness`] object from an array of byte-arrays
+    pub fn from_vec(vec: Vec<Vec<u8>>) -> Self {
```
Nit: why not `impl From<Vec<Vec<u8>> for Witness` instead?

And probably `pub fn from<T: AsRef<[u8]>>(bytes: &[T])` will be better.

apoelstra replied:
Because `From` is for cheap/trivial conversions. This is neither.

Kixunil replied:
> `From` is for cheap/trivial conversions

Not really, `impl From<&str> for String` exists for instance and that conversion is not usually considered "cheap".

@dr-orlovsky I consider `AsRef<[u8]>` a footgun since it would accept strings without explicit `as_bytes()`

dr-orlovsky commented on src/blockdata/witness.rs:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    /// Returns a struct implementing [`Iterator`]
+    pub fn iter(&self) -> Iter {
```
I think the name here is confusing: it is not clear do we iterate over witness stack elements - or witness byte sequence. So I will prefer `fn byte_iter(&self)`

apoelstra replied:
We iterate over witness stack elements. It wouldn't be meaningful to iterate over bytes. I don't think there is any ambiguity here.

dr-orlovsky replied:
From what I see in the code we do iterate over bytes here: `self.content.iter()` where `content: Vec<u8>` will iterate over bytes.

RCasatta replied:
`self.content` is used internally to provide the witness stack elements iterator via `iter()`

Kixunil replied:
I agree with @dr-orlovsky here, `chars()` and `bytes()` exist on `&str` for this reason.

apoelstra replied:
It is clearly meaningful to iterate over either the characters or bytes of a string. It is not clear to me what it means to iterate over the bytes of a witness, or even in what order they would be returned.

Kixunil replied:
OK, if it seems completely obvious that it means stack elements fine for me. I'd just like the docs to be more descriptive.

Just note that non-obvious iterators can be a footgun. I once got burned by `&std::path::Path` implementing `IntoIterator` (Can you guess what it iterates over without looking it up?)

dr-orlovsky reviewed (approved):
utACK 106acdc3ac54372c4cc0a4a011a2ab8848159643

I have several notes, but I do not think it worth re-reviewing this PR due to them - they may be addressed in a follow-up PR.
